jax/_src/api.py

```diff
@@ -467,9 +467,100 @@ def get_device_info():
                              cache=_cpp_jit_cache)
   f_jitted = wraps(fun)(cpp_jitted_f)
 
+  f_jitted.lower = partial(_lower,
+                           fun, static_argnums, static_argnames, device,
+                           backend, donate_argnums, inline)
+
   return f_jitted
 
 
+class JitOptions(NamedTuple):
+  static_argnums: Union[int, Iterable[int], None]
+  static_argnames: Union[str, Iterable[str], None]
+  donate_argnums: Union[int, Iterable[int]]
+
+
+class JitLowered:
+  __slots__ = ['computation', 'options', 'source_fun', 'in_tree', 'out_tree']
+
+  computation: xla.XlaComputation
+  options: JitOptions
+  source_fun: Callable
+  in_tree: PyTreeDef
+  out_tree: PyTreeDef
+
+  def __init__(self, computation, options, source_fun, in_tree, out_tree):
+    self.computation = computation
+    self.options = options
+    self.source_fun = source_fun
+    self.in_tree = in_tree
+    self.out_tree = out_tree
+
+  def _xla_computation(self):
+    # TODO(frostig): finalize API. For now, return the underlying
+    # computation directly via this method.
+    if self.computation.is_trivial():
+      raise ValueError('A trivial computation has no HLO')
+    return self.computation.hlo
+
+  def compile(self):
+    return JitCompiled(
+        self.computation.compile(), self.options, self.source_fun,
+        self.in_tree, self.out_tree)
+
+
+class JitCompiled:
+  __slots__ = ['computation', 'options', 'source_fun', 'in_tree', 'out_tree']
+
+  computation: xla.XlaComputation
+  options: JitOptions
+  source_fun: Callable
+  in_tree: PyTreeDef
+  out_tree: PyTreeDef
+
+  def __init__(self, computation, options, source_fun, in_tree, out_tree):
+    self.computation = computation
+    self.options = options
+    self.source_fun = source_fun
+    self.in_tree = in_tree
+    self.out_tree = out_tree
+
+  def _xla_executable(self):
+    # TODO(frostig): finalize API. For now, return the underlying
+    # executable directly via this method.
+    if self.computation.is_trivial():
+      raise ValueError('A trivial computation has no executable')
+    return self.computation.xla_executable
+
+  def __call__(self, *args, **kwargs):
+    _, in_tree, args_flat, _ = _prepare_jit(
+        self.source_fun, self.options.static_argnums,
+        self.options.static_argnames, self.options.donate_argnums,
+        args, kwargs)
+    if in_tree != self.in_tree:
+      # TODO(frostig): provide more info about the source function
+      raise TypeError(
+          f'function compiled for {self.in_tree}, called with {in_tree}')
+    out_flat = self.computation.call(*args_flat)
+    return tree_unflatten(self.out_tree, out_flat)
+
+
+def _lower(fun, static_argnums, static_argnames, device, backend,
```

Comment:
Wouldn't it make sense to use `_lower` on the regular execution path too? That's what I do with `xmap` and `pjit` (though `jit` has the added difficult of having a C++ path, which I don't know too much about).